proposals/low-level-struct-improvements.md

```diff
@@ -939,17 +969,133 @@ This will work but is likely going to result in unnecessary code generation. One
+        // prevents us from getting here.
+        return local;
+    }
+}
+```
+
+Essentially it would mean all instance method invocations on *mutable* `ref struct` locals would be illegal unless the local was further marked as `scoped`. The rules have to consider the case where fields were ref re-assigned to other fields in `this`. A `readonly ref struct` doesn't have this problem because the `readonly` nature prevents ref re-assignment. Still this would be a significant back compat breaking change as it would impact virtually every existing mutable `ref struct`. 
+
+Some thought was given to the idea of having `this` have different defaults based on the type of `struct` or member. For example:
+
+ - `this` as `ref`: `struct`, `readonly ref struct` or `readonly member`
+ - `this` as `scoped ref`: `ref struct`
+
+ This minimizes compat breaks and maximizes flexibility but at the cost of complicating the story for customers. It also doesn't fully solve the problem because future features, like safe `fixed` buffers, require that a mutable `ref struct` have `ref` returns for fields which don't work by this design alone as it would fall into the `scoped ref` category. 
+
+**Decision** Keep `this` as `scoped ref`
+
+### ref fields to ref struct
+This feature opens up a new set of ref safety rules because it allows for a `ref` field to refer to a `ref struct`. This generic nature of `ByReference<T>` meant that up until now the runtime could not have such a construct. As a result all of our rules are written under the assumption this is not possible. The `ref` field feature is largely not about making new rules but codifying the existing rules in our system. Allowing `ref` fields to `ref struct` requires us to codify new rules because there are several new scenarios to consider.
+
+The first is that a `readonly ref` is now capable of storing `ref` state. For example:
+
+```c#
+readonly ref Container
+{
+    readonly ref Span<int> Span;
+
+    void Store(Span<int> span)
+    {
+        Span = span;
+    }
+}
+```
+
+This means when thinking about method arguments must match rules we must consider `readonly ref T` is potential method output sources when `T` potentially has a `ref` field to a `ref struct`.
+
+The second issue is language must consider a new type of escape scope: *ref-field-safe-to-escape*. All `ref struct` which transitively contain a `ref` field have another escape scope representing the value(s) in the `ref` field(s). In the case of multiple `ref` fields they can be collectively tracked as a single value. The default value for this for parameters is *calling method*. 
+
+```c#
+ref struct Nested
+{
+    ref Span<int> Span;
+}
+
+Span<int> M(ref Nested nested) => nested.Span;
+```
+
+This value is not related to the escape scope of the container; that is as it the container scope gets smaller it has no impact on the *ref-field-safe-to-escape* of the `ref` field values. 
```

Comment:
>This value is not related to the escape scope of the container; that is as it the container scope gets smaller it has no impact on the *ref-field-safe-to-escape* of the `ref` field values. [](http://example.com/codeflow?start=0&length=188)

Is this correct? Can't we say the *ref-field-safe-to-escape* scope is no smaller than the container scope?

Reply:
> Is this correct?

The scenario I'm trying to get at is the following: 

```c#
// The RFSTE and STE of `nested` is *calling method*
void M(Nested nested)
{
  // The STE is *current method* 
  // The RFST is still *calling method* 
  scoped Nested local = nested; 

}
```

That is correct, otherwise it allows for us to do actions like `local.Span = stackalloc`. If I can make this clearer happy to do so. 

> Can't we say the ref-field-safe-to-escape scope is no smaller than the container scope?

Can add that too.